### ISSUE TITLE
Jobs can leak disk space

### ISSUE THREAD

**Issue by sjezewski** [pachyderm/pachyderm#1388]
Right now if a job fails in a way that brings down the pod (e.g. a panic, or if k8s evicts the pod), our GC code to cleanup the disk doesn't get run and we can leak disk space.

Personally I've seen this happen in the eviction case.

This can get dangerous or even pathological if it feeds back enough. E.g. if a job fails because there wasn't enough room on the node to write to `/pfs/out` then gets restarted again. If a single job consistently writes more data than there is space on the node, then this can take down the whole cluster (eventually everything gets evicted because of disk space).

To diagnose, ssh into a node and run:

```
sudo du -h -d 5 / | sort -h
```

And if you see most of the space consumed under `/pach-job-data/` or `/var/pachyderm/pach-job-data` (depending on your pachd version) then this is a smoking gun. (Note: running jobs can legitimately use this space ... but if you're disk is full and its mostly that folder ... this is likely the cause)

---

First some terms:

```
N = # of nodes
P = parallelism
M = total disk output from a single run of a job (across all pods / nodes). Also this should be the maximum possible across jobs of this type
V = volume size per node
```

For now, we recommend:

- making the pipeline have `lazy` input (so that you minimize disk usage from data input) ... if the total number of input files is less than ~10k (more specifically 10k per node depending on the parallelism)
- upping the number of nodes `N` (and parallelism to 'use' all the nodes), and your VM volume sizes `V` so that `M / N < V`

e.g. If I had a job that needed to process 120Gb of data, and I currently have N=4, V=20, P=4, that won't work. Each pod will try and write 30Gb of data, and that will flood the disk. In this case I could up V=40 and it would be ok.

We're looking into other ways to make the tooling more resilient to prevent this, and more descriptive when disk usage is organically exhausted so that we can message users to this effect (w steps for recourse). For now, this is a beacon issue, as the job execution model will be changing in `1.4.`

**Comment by derekchiang** [comment]
Removing `high priority` because this is less of an issue in `1.4`

**Comment by derekchiang** [comment]
I think this is fixed in #1772
